### PR TITLE
allow IO objects as file target

### PR DESCRIPTION
**Pull request by akeeman**
The library is able to help with downloading data to a local file. Sometimes the target is not local, however. It would be great to allow not only writing to local files, but any IO object, which may represent remote files.

This is basically implemented by adding `io.RawIOBase` and `io.BufferedIOBase` to the allowed target types.

### CODE REVIEW

File: cdsapi/api.py [hunk 1]
```diff
@@ -101,60 +121,58 @@ def _download(self, url, size, target):
...
-                    total=size,
-                    unit_scale=True,
-                    unit_divisor=1024,
-                    unit="B",
-                    disable=not self.progress,
-                    leave=False,
-                ) as pbar:
-                    pbar.update(total)
-                    with open(target, mode) as f:
+        with _open_target(target) as f:
```
Comment (akeeman) [inline cdsapi/api.py:129]: moved opening the file to here, so we may open it once.

File: cdsapi/api.py [hunk 2]
```diff
@@ -101,60 +121,58 @@ def _download(self, url, size, target):
...
-                sleep = self.sleep_max
-            headers = {"Range": "bytes=%d-" % total}
-            tries += 1
-            self.warning("Resuming download at byte %s" % (total,))
+                self.error(
+                    "Download incomplete, downloaded %s byte(s) out of %s" % (total, size)
+                )
+                self.warning("Sleeping %s seconds" % (sleep,))
+                time.sleep(sleep)
+                total = f.tell()
```
Comment (akeeman) [inline cdsapi/api.py:169]: we may check the position of the write pointer now to find the total amount of bytes written.

File: cdsapi/api.py [hunk 3]
```diff
@@ -101,60 +121,58 @@ def _download(self, url, size, target):
...
-                break
+                if total >= size:
+                    break
 
-            self.error(
-                "Download incomplete, downloaded %s byte(s) out of %s" % (total, size)
-            )
-            self.warning("Sleeping %s seconds" % (sleep,))
-            time.sleep(sleep)
-            mode = "ab"
```
Comment (akeeman) [inline cdsapi/api.py:150]: file pointer stays open, so no need to change modes.

File: cdsapi/api.py [hunk 4]
```diff
@@ -101,60 +121,58 @@ def _download(self, url, size, target):
         self.info("Downloading %s to %s (%s)", url, target, bytes_to_string(size))
         start = time.time()
 
-        mode = "wb"
```
Comment (akeeman) [inline cdsapi/api.py:104]: file pointer stays open, so no need to change modes.

File: cdsapi/api.py [hunk 5]
```diff
@@ -101,60 +121,58 @@ def _download(self, url, size, target):
...
                                 f.write(chunk)
                                 total += len(chunk)
                                 pbar.update(len(chunk))
 
-            except requests.exceptions.ConnectionError as e:
-                self.error("Download interupted: %s" % (e,))
-            finally:
-                r.close()
+                except requests.exceptions.ConnectionError as e:
+                    self.error("Download interupted: %s" % (e,))
```
Comment (akeeman) [inline cdsapi/api.py:157]: fixed typo: `interupted` -> `interrupted`

File: tests/requirements.txt [hunk 6]
```diff
@@ -7,3 +7,4 @@ pytest-pep8
 pytest-runner
 pytest
 requests
+requests-mock
```
Comment (akeeman) [inline tests/requirements.txt:10]: let's add some mocked http calls for non-end-to-end tests

File: cdsapi/api.py [hunk 7]
```diff
@@ -6,13 +6,15 @@
 # granted to it by virtue of its status as an intergovernmental organisation nor
 # does it submit to any jurisdiction.
 
-from __future__ import absolute_import, division, print_function, unicode_literals
+from __future__ import annotations, absolute_import, division, print_function, unicode_literals
```
Comment (akeeman) [inline cdsapi/api.py:9]: enable postponed evaluation of annotations

File: cdsapi/api.py [hunk 8]
```diff
@@ -62,6 +64,24 @@ def toJSON(obj):
...
+def _open_target(target: str | bytes | os.PathLike | io.RawIOBase | io.BufferedIOBase) -> IO[bytes]:
+    """Creates a BinaryIO context manager given a certain destination file target."""
+    ERR_MSG = "Invalid value for target. Use None, path string or bytes, or a file-like object."
+
+    if isinstance(target, (str, bytes, os.PathLike)):
+        file = open(target, "wb")
+    elif isinstance(target, (io.RawIOBase, io.BufferedIOBase)):
+        try:
+            file = target
+            file.seek(0)
```
Comment (akeeman) [inline cdsapi/api.py:76]: make sure we start to write at the start of the file for accurate download size counting

**Review by gbiavati — APPROVED** [review]

**Comment by niowniow** [comment]
Thanks @akeeman ! Is there a chance that this could be merged by a maintainer soon?

**Comment by akeeman** [comment]
This pr is useless since the migration to https://github.com/ecmwf/multiurl/issues, which also does not support writing to IO objects.

**Pull request closed without merging**